### PR TITLE
opt: generate more non-NULL constraints, extract them to NotNullCols

### PR DESCRIPTION
**Pull request by RaduBerinde**
#### opt: generate more not-NULL constraints

Generate not-NULL constraints from expressions like `a >
something-non-const` or `a = b`. The latter is especially useful for
inner joins.

Release note: None

#### opt: extract not-null columns from filter constraints

We use constraints to populate NotNullCols for SelectOp.

Release note: None

### CODE REVIEW

**Comment by andy-kimball** [comment]
:lgtm:, mostly nits.

---

Review status: 0 of 50 files reviewed at latest revision, all discussions resolved.

---

*[pkg/sql/opt/constraint/constraint_set.go, line 263 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25423#-LCG641yEDwrBTrFWC6o:-LCG641yEDwrBTrFWC6p:biyklvz) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c79053af94b16f91c7a94558b675e461905ac79/pkg/sql/opt/constraint/constraint_set.go#L263)):*
> ```Go
> 		firstCol := c.Columns.Get(0)
> 
> 		// Deduce a constraint on the first column.
> ```

What happens if the column is a `DESC` column?

---

*[pkg/sql/opt/constraint/constraint_set.go, line 271 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25423#-LCG6QzF8jzvS5A5elI4:-LCG6QzF8jzvS5A5elI5:b-i9cwh4) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c79053af94b16f91c7a94558b675e461905ac79/pkg/sql/opt/constraint/constraint_set.go#L271)):*
> ```Go
> 			res.Add(int(firstCol))
> 		}
> 		// TODO(radu): try to deduce more not-null columns from multi-column
> ```

NIT: I think we should normalize that particular constraint to be 3 diff constraints on the individual columns, in this particular case.

---

*[pkg/sql/opt/memo/logical_props_builder.go, line 148 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25423#-LCG7K4w9n8Q-MxDZR-W:-LCG7K4w9n8Q-MxDZR-X:b6i8zwh) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c79053af94b16f91c7a94558b675e461905ac79/pkg/sql/opt/memo/logical_props_builder.go#L148)):*
> ```Go
> 	b.sb.buildSelect(ev.Child(1), &inputProps.Stats)
> 
> 	if filterProps.Constraints != nil && filterProps.Constraints != constraint.Unconstrained {
> ```

Move this above the stats initialization (and before the outercols init), b/c that oftentimes expects the other props to already be set. I also try to maintain roughly the same ordering for setting props across the different ops.

---

*[pkg/sql/opt/memo/testdata/logprops/select, line 93 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25423#-LCG7u546SJK9B1oDw9Q:-LCG7u566bW0VEjs7WEn:b9nopjw) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c79053af94b16f91c7a94558b675e461905ac79/pkg/sql/opt/memo/testdata/logprops/select#L93)):*
> ```
> # Reduce min cardinality.
> build
> SELECT COUNT(*) FROM xy HAVING COUNT(*) = 5
> ```

NIT: I'd modify and/or add a more direct test case for the new code.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25423#-:-LCG8fCE4g7qAdd7i7Nz:b-wze63b)*

**Comment by RaduBerinde** [comment]
Review status: 0 of 50 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/sql/opt/constraint/constraint_set.go, line 263 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25423#-LCG641yEDwrBTrFWC6o:-LCHA6HWAjYstadA-MHQ:b-vhvrnt) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c79053af94b16f91c7a94558b675e461905ac79/pkg/sql/opt/constraint/constraint_set.go#L263)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

What happens if the column is a `DESC` column?
</blockquote></details>

Good point. I skipped the constraint added a TODO (we don't really need it currently)

---

*[pkg/sql/opt/constraint/constraint_set.go, line 271 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25423#-LCG6QzF8jzvS5A5elI4:-LCH9GxWC83mqmEVYfYJ:b-juiuhk) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c79053af94b16f91c7a94558b675e461905ac79/pkg/sql/opt/constraint/constraint_set.go#L271)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

NIT: I think we should normalize that particular constraint to be 3 diff constraints on the individual columns, in this particular case.
</blockquote></details>

I agree, definitely. But the TODO still holds for cases like `[/1/1/1 - /1/1/2], [/3/3/3 - /3/3/4]` (much less important though)

---

*[pkg/sql/opt/memo/logical_props_builder.go, line 148 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25423#-LCG7K4w9n8Q-MxDZR-W:-LCHB2jjEWU5Za88p0b0:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c79053af94b16f91c7a94558b675e461905ac79/pkg/sql/opt/memo/logical_props_builder.go#L148)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Move this above the stats initialization (and before the outercols init), b/c that oftentimes expects the other props to already be set. I also try to maintain roughly the same ordering for setting props across the different ops.
</blockquote></details>

Done.

---

*[pkg/sql/opt/memo/testdata/logprops/select, line 93 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25423#-LCG7u546SJK9B1oDw9Q:-LCHC--k-GOkqLrnokf9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c79053af94b16f91c7a94558b675e461905ac79/pkg/sql/opt/memo/testdata/logprops/select#L93)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

NIT: I'd modify and/or add a more direct test case for the new code.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25423)*

**Comment by rytaft** [comment]
First commit LGTM, still looking at the second

---

Reviewed 7 of 24 files at r1, 43 of 43 files at r3.
Review status: 8 of 50 files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25423#-:-LCHVpi37uHsaFQ3wPVR:b-pzaan5)*

**Comment by rytaft** [comment]
:lgtm:

---

Reviewed 42 of 42 files at r4.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/sql/opt/constraint/constraint_set.go, line 265 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/25423#-LCUxK2HCb-BeN_sVpAd:-LCUxK2HCb-BeN_sVpAe:bijgxzd) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c878e9438726e769347cda5b98ef88dae731b2c/pkg/sql/opt/constraint/constraint_set.go#L265)):*
> ```Go
> 			// TODO(radu): handle descending columns. Currently this is not useful
> 			// because this function operates on constraints generated by the memo
> 			// constraint builder which don't use descending columns.
> ```

[nit] don't -> doesn't

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25423#-:-LCV0S9CDIRQzLj5tWw7:bnfp4nl)*

**Comment by RaduBerinde** [comment]
Review status: 3 of 51 files reviewed at latest revision, 5 unresolved discussions.

---

*[pkg/sql/opt/constraint/constraint_set.go, line 265 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/25423#-LCUxK2HCb-BeN_sVpAd:-LCV4f5PBLhQQHDG79l2:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c878e9438726e769347cda5b98ef88dae731b2c/pkg/sql/opt/constraint/constraint_set.go#L265)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

[nit] don't -> doesn't
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25423)*
